scylla-cql/src/types/deserialize/frame_slice.rs

```diff
@@ -0,0 +1,188 @@
+use bytes::Bytes;
+
+use crate::frame::frame_errors::ParseError;
+use crate::frame::types;
+
+/// A reference to a part of the frame.
+//
+// # Design justification
+//
+// ## Why we need a borrowed type
+//
+// The reason why we need to store a borrowed slice is that we want a lifetime that is longer than one obtained
+// when coercing Bytes to a slice in the body of a function. That is, we want to allow deserializing types
+// that borrow from the frame, which resides in QueryResult.
+// Consider a function with the signature:
+//
+// fn fun(b: Bytes) { ... }
+//
+// This function cannot return a type that borrows from the frame, because any slice created from `b`
+// inside `fun` cannot escape `fun`.
+// Conversely, if a function has signature:
+//
+// fn fun(s: &'frame [u8]) { ... }
+//
+// then it can happily return types with lifetime 'frame.
+//
+// ## Why we need the full frame
+//
+// We don't. We only need to be able to return Bytes encompassing our subslice. However, the design choice
+// was made to only store a reference to the original Bytes object residing in QueryResult, so that we avoid
+// cloning Bytes when performing subslicing on FrameSlice. We delay the Bytes cloning, normally a moderately
+// expensive operation involving cloning an Arc, up until it is really needed.
+//
+// ## Why not different design
+//
+//      - why not a &'frame [u8] only? Because we want to enable deserializing types containing owned Bytes, too.
+//      - why not a Bytes only? Because we need to propagate the 'frame lifetime.
+//      - why not a &'frame Bytes only? Because we want to somehow represent subslices, and subslicing
+//        &'frame Bytes return Bytes, not &'frame Bytes.
+#[derive(Clone, Copy, Debug)]
+pub struct FrameSlice<'frame> {
+    // The actual subslice represented by this FrameSlice.
+    frame_subslice: &'frame [u8],
+
+    // Reference to the original Bytes object that this FrameSlice is derived
+    // from. It is used to convert the `mem` slice into a fully blown Bytes
+    // object via Bytes::slice_ref method.
+    original_frame: &'frame Bytes,
+}
+
+static EMPTY_BYTES: Bytes = Bytes::new();
+
+impl<'frame> FrameSlice<'frame> {
+    /// Creates a new FrameSlice from a reference of a Bytes object.
+    ///
+    /// This method is exposed to allow writing deserialization tests
+    /// for custom types.
+    #[inline]
+    pub fn new(frame: &'frame Bytes) -> Self {
+        Self {
+            frame_subslice: frame,
+            original_frame: frame,
+        }
+    }
+
+    /// Creates a new FrameSlice that refers to a subslice of a given Bytes object.
+    #[inline]
+    pub fn new_subslice(mem: &'frame [u8], frame: &'frame Bytes) -> Self {
+        Self {
+            frame_subslice: mem,
+            original_frame: frame,
+        }
+    }
```

Comment:
> Other methods guarantee that `frame_subslice` points to the same slice as `original_frame`, but this one allows creation of ill-formed objects.

This is not necessarily true. Consider `as_slice_mut`. The caller of it can even replace the slice with an arbitrary slice satisfying the given lifetime - e.g. using a `'static` slice.

> Modifying `FrameSlice` using `as_slice_mut` seems less error-prone to me than creating new slice and then a new object using this slice.

Mostly yes.

Pushed a new version with `new_subslice` ditched.